### PR TITLE
Riedl

### PR DESCRIPTION
**Pull request by jriedl**

### CODE REVIEW

File: cpp/journal_publishing_processor.cc [hunk 1]
```diff
@@ -124,7 +129,26 @@ bool ExtractText(XMLParser * const xml_parser, const std::string &text_opening_t
 }
 
 
-void ProcessDocument(const bool normalise_only, const std::string &input_file_path, XMLParser * const xml_parser,
+void ExtractPDFFulltext(const bool force_ocr, const std::string &fulltext_location, std::string * const full_text) {
+    if (not StringUtil::EndsWith(fulltext_location, ".pdf", true))
```
Comment (ruschein) [inline cpp/journal_publishing_processor.cc]: Please always add a comment in front of cryptic constants like "true" or "false".  In this case I'd write /* ignore_case = */.

File: cpp/journal_publishing_processor.cc [hunk 2]
```diff
@@ -124,7 +129,26 @@ bool ExtractText(XMLParser * const xml_parser, const std::string &text_opening_t
...
+    if (not force_ocr) {
+        if (not PdfUtil::PdfDocContainsNoText(pdf_document))
+            PdfUtil::ExtractText(pdf_document, full_text);
+        else
+            LOG_ERROR("Apparently no text in \"" + fulltext_location + "\"");
+    } else {
+        if (not PdfUtil::GetOCRedTextFromPDF(fulltext_location, "eng+grc+heb", full_text, 120))
+            LOG_ERROR("Could not extract text from \"" + fulltext_location + "\"");
+    }
+
```
Comment (ruschein) [inline cpp/journal_publishing_processor.cc]: Extraneous black line.

File: cpp/journal_publishing_processor.cc [hunk 3]
```diff
@@ -124,7 +129,26 @@ bool ExtractText(XMLParser * const xml_parser, const std::string &text_opening_t
...
+        LOG_ERROR("Don't know how to handle file \"" + fulltext_location + "\"");
+    std::string pdf_document;
+    if (not FileUtil::ReadString(fulltext_location, &pdf_document))
+        LOG_ERROR("Could not read \"" + fulltext_location + "\"");
+    if (not force_ocr) {
+        if (not PdfUtil::PdfDocContainsNoText(pdf_document))
+            PdfUtil::ExtractText(pdf_document, full_text);
+        else
+            LOG_ERROR("Apparently no text in \"" + fulltext_location + "\"");
+    } else {
```
Comment (ruschein) [inline cpp/journal_publishing_processor.cc]: Turn that into an "else if", please.

File: cronjobs/fetch_marc_updates.py [hunk 4]
```diff
@@ -304,12 +304,22 @@ def DownloadData(config, section, ftp, download_cutoff_date, msg):
...
 def DownloadCompleteData(config, ftp, download_cutoff_date, msg):
     downloaded_files = DownloadData(config, "Kompletter Abzug", ftp, download_cutoff_date, msg)
-    if len(downloaded_files) == 1:
-        return downloaded_files[0]
-    elif len(downloaded_files) == 0:
-        return None
+    complete_filename_pattern = config.get("Kompletter Abzug", "filename_pattern")
+    if not NeedsBothInstances(re.compile(complete_filename_pattern)):
+        if len(downloaded_files) == 1:
+            return downloaded_files
```
[the remark this answers is not in the document]
Comment (jriedl) [reply cronjobs/fetch_marc_updates.py]: The complete dump now logically consists of two files (with and without local data), so array approach is needed.

**Pull request merged**